src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleMetadata.cs

```diff
@@ -14,42 +14,31 @@ public static partial class JsonSerializer
         internal static readonly JsonEncodedText s_metadataRef = JsonEncodedText.Encode("$ref", encoder: null);
         internal static readonly JsonEncodedText s_metadataValues = JsonEncodedText.Encode("$values", encoder: null);
 
-        internal static MetadataPropertyName GetResolvedReferenceHandling(
-            JsonConverter converter,
-            object value,
-            ref WriteStack state,
-            out string? referenceId)
-        {
-            if (!converter.CanHaveIdMetadata || converter.TypeToConvert.IsValueType)
-            {
-                referenceId = default;
-                return MetadataPropertyName.NoMetadata;
-            }
-
-            if (state.ReferenceResolver.TryGetOrAddReferenceOnSerialize(value, out referenceId))
-            {
-                return MetadataPropertyName.Ref;
-            }
-
-            return MetadataPropertyName.Id;
-        }
-
         internal static MetadataPropertyName WriteReferenceForObject(
             JsonConverter jsonConverter,
             object currentValue,
             ref WriteStack state,
             Utf8JsonWriter writer)
         {
-            MetadataPropertyName metadataToWrite = GetResolvedReferenceHandling(jsonConverter, currentValue, ref state, out string? referenceId);
+            MetadataPropertyName metadataToWrite;
```

Comment:
Was the motivation here only to remove the `!` in `referenceId!` or also other reasons (perf?)

Reply:
That and perf (avoiding unnecessary/duplicate checks on the enum).